text/0122-sdk-hub-scope-merge.md

```diff
@@ -0,0 +1,475 @@
+- Start Date: 2023-11-07
+- RFC Type: feature
+- RFC PR: https://github.com/getsentry/rfcs/pull/122
+- RFC Status: draft
+
+# Summary
+
+We want to streamline Hub & Scope usage for the SDK.
+
+# Motivation
+
+Generally, Hubs and Scopes are a very Sentry-specific concept, and can be a bit hard to explain.
+Also, how Hub & Scope forking behaves is currently a bit tricky, and does not play well e.g. with how OpenTelemetry contexts work.
+
+This RFC aims to streamline this by merging the concepts of Hub and Scope into a singular Scope concept.
+
+It also proposes the new concepts of global & isolated scopes.
+
+# Background
+
+Hubs and Scopes have been part of the unified SDK API for a long time.
+While they have generally served us well, the shortcomings of the current systems have become apparent over time.
+Apart from generally being unclear to users - what is applied when, why, how? - there are also problems with the way we currently instruct users to add event data.
+
+A major reason for this change is also to better align with OpenTelemetry. In OpenTelemetry (OTEL), the equivalent to our scope is the "Context". However, the Context is immutable in OTEL, and works via copy-on-write - so each time you want to change _anything_ on the context, you fork it. This means that you have much more forking than you have in Sentry currently, and also since you always fork a context when you change the active span (as the span is set on the context, and updating it means forking a context), you can quickly get into deeply nested context situations due to auto instrumentation. 
```

Comment:
This RFC doesn't seem to mention where to store the current span. While we might no longer have a Sentry span after completely switching over to OTEL, there'll likely be intermediate steps as well as some SDKs that can't yet move to pure OTEL spans.

Where we could put it:
- Isolation span
	- I don't think that's a good idea as
		- It's a potential footgun allowing users to hide the current span set on isolation scope by the SDK by setting a span on the current span
		- When a span ends we have to restore the previous span on the isolation scope
- Current Span
	- Since we're forking current scope anytime a new span is created anyways this is a good candidate
	- A drawback here is that it requires a more complex copy-on-write implementation that's per property instead of simply copying the whole scope. It seems the number one reason for copying scope will be creating a new span.
- Execution Context
	- This would keep current span separate from current scope and thus not cause as many current scope copies